### ISSUE TITLE
poor precision of unet backbone with resnet50 when deal with multi-classes segmentation

### ISSUE THREAD

**Issue by yasohasakii** [qubvel/segmentation_models#276]
Hi Pavel! I have some trouble when using your segmentation models with multi-classes segmentation images. The images are medical histopathlogical image and classes are imbalanced which means there're lots of negative instances and little of positive instances. I labeled negative samples with green `(0,255,0)` mask and positive samples with red `(255,0,0)` and blue`(0,0,255)`. However, the result is terrible like below 
![predict result is the right one, the middle one is ground truth](https://github.com/yasohasakii/unet-segmentation/blob/yasohasakii-multiclass-segmentation/test/05914499c3484d51988751ace447b639.png)
I'd like to put my codes to figure out the problem
>optim = keras.optimizers.Adam(1e-3)
>dice_loss = sm.losses.DiceLoss()
>focal_loss = sm.losses.CategoricalFocalLoss()
>total_loss = dice_loss + (1. * focal_loss)
>metrics = [sm.metrics.IOUScore(threshold=0.5,per_image=True)\ 
>    ,sm.metrics.FScore(threshold=0.5,per_image=True)]
> model = sm.Unet('resnet50' ,classes=3 ,activation='softmax',input_shape = (512,512,3))
>model.compile(optim ,loss=total_loss ,metrics=metrics)

**Comment by JordanMakesMaps** [comment]
Based on the GT data, it looks like you have some imbalance between your negative and positives. Might want to look at using different loss functions, and/or weighting the classes accordingly. It looks like `DiceLoss` can include weighted classes so you could try that first:

``` python

def __init__(self, beta=1, class_weights=None, class_indexes=None, per_image=False, smooth=SMOOTH)

```
I am curious though, @qubvel, if someone wanted to make their own instance of say, cce_dice_loss, and they wanted to use weights for their classes due to imbalances, which is the right way of doing it? Because both allow for `class_weights` to be included.

```python
#example weights and index list
class_weights = [.25, .25, .25, .1] 
class_index = [0, 1, 2, 3, 4]

# method 1
dice = DiceLoss(class_weights = class_weights, class_indexes = class_index)
cce = CategoricalCELoss(class_weights = class_weights, class_indexes = class_index)
loss_function = dice + cce

# or

# method 2
dice = DiceLoss(class_weights = class_weights, class_indexes = class_index)
cce = CategoricalCELoss()
loss_function = dice + cce
```

And regarding @yasohasakii's situation, would the right way be....?
```python

#example weights and index list
class_weights = [.25, .25, .25, .1] 
class_index = [0, 1, 2, 3, 4]

dice = DiceLoss(class_weights = class_weights, class_indexes = class_index)
focal = CategoricalFocalLoss(class_indexes = class_index)
loss_function = dice + focal
```

**Comment by yasohasakii** [comment]
> Based on the GT data, it looks like you have some imbalance between your negative and positives. Might want to look at using different loss functions, and/or weighting the classes accordingly. It looks like `DiceLoss` can include weighted classes so you could try that first:
> 
> ```python
> def __init__(self, beta=1, class_weights=None, class_indexes=None, per_image=False, smooth=SMOOTH)
> ```
> 
> I am curious though, @qubvel, if someone wanted to make their own instance of say, cce_dice_loss, and they wanted to use weights for their classes due to imbalances, which is the right way of doing it? Because both allow for `class_weights` to be included.
> 
> ```python
> #example weights and index list
> class_weights = [.25, .25, .25, .1] 
> class_index = [0, 1, 2, 3, 4]
> 
> # method 1
> dice = DiceLoss(class_weights = class_weights, class_indexes = class_index)
> cce = CategoricalCELoss(class_weights = class_weights, class_indexes = class_index)
> loss_function = dice + cce
> 
> # or
> 
> # method 2
> dice = DiceLoss(class_weights = class_weights, class_indexes = class_index)
> cce = CategoricalCELoss()
> loss_function = dice + cce
> ```
> 
> And regarding @yasohasakii's situation, would the right way be....?
> 
> ```python
> #example weights and index list
> class_weights = [.25, .25, .25, .1] 
> class_index = [0, 1, 2, 3, 4]
> 
> dice = DiceLoss(class_weights = class_weights, class_indexes = class_index)
> focal = CategoricalFocalLoss(class_indexes = class_index)
> loss_function = dice + focal
> ```

Thanks ! I'd try and reply the result soon.

**Comment by yasohasakii** [comment]
Found that `class_weights` is complicate to count, some blogs recommend cross validation can get class weights. And I believe duplicate sampling and data augment method combine can also solve the imbalance problem in better way.
